enhancements/monitoring/integration-with-topology-modes.md

```diff
@@ -0,0 +1,169 @@
+
+Otherwise CMO deploys the operands with their default number of replicas to
+ensure high-availability:
+
+* openshift-monitoring/prometheus-k8s (StatefulSet): 2 replicas
+* openshift-monitoring/alertmanager-main (StatefulSet): 3 replicas
+* openshift-monitoring/thanos-querier (Deployment): 2 replicas
+* openshift-monitoring/prometheus-adapter (Deployment): 2 replicas
+* openshift-user-workload-monitoring/prometheus-user-workload (StatefulSet): 2 replicas
+* openshift-user-workload-monitoring/thanos-ruler (StatefulSet): 2 replicas
+
+By default, CMO deploys no operand on control nodes thus it should not look at
+the `controlPlaneTopology` field.
+
+### Risks and Mitigations
+
+In HA mode, the Prometheus instances are spread on different nodes using
+anti-affinity and they work completely indepently one from another (there's no
+data replication between them). Services consuming the metrics (like the
+OpenShift console or Grafana) use the Thanos querier API which aggregates and
+deduplicates data from both Prometheus instances. This setup provides
+redundancy in term of metrics scraping, rule evaluations and querying as long
+as one of the Prometheus instances is still up.
+
+The Thanos ruler instances of user workload monitoring work the same as
+Prometheus except they only deal with rule evaluations.
+
+The 3 Alertmanager instances receive alerts from both Prometheus instances and
+as long as one instance is up and running, the notifications should flow to the
+receivers. Alertmanager instances also replicate silences and notification logs
+between themselves to prevent duplicate notifications.
+
+The other monitoring components running with multiple replicas in HA mode
+(thanos-querier and prometheus-adapter) are stateless. Running multiple
+replicas only ensures that their services remain available as long as 1
+instance is up.
+
+By definition, the monitoring stack wouldn't be highly-available in the
+`SingleReplica` mode. One recommendation would be that cluster admins configure
+Prometheus and Alertmanager with persistent storage to ensure that existing
+metrics, silences and notification logs are retained upon pod recreation.
+
+## Design Details
+
+### Test Plan
+
+We will setup a CI job that runs the `extended` end-to-end test suite from
+[OpenShift Origin](https://github.com/openshift/origin) in a single-node
+deployment mode. The test suite has already been extended to verify that
+operators honor the `SingleReplica` value (see
+[openshift/origin#25812](https://github.com/openshift/origin/pull/25812) and
+[openshift/origin/pull/25885](https://github.com/openshift/origin/pull/25885)).
+
+As soon as the implementation in CMO is merged, we will adjust the single-node
+topology
+[tests](https://github.com/openshift/origin/blob/master/test/extended/single_node/topology.go)
+to remove the exception in place for the monitoring components.
+
+Because the topology mode isn't exposed into the CMO configuration, we don't plan
```

Comment:
I'm not sure I understand this statement. Does it refer to the configuration of the CMO itself, rather than the operands?

Reply:
I will clarify but it refers to the configuration that CMO exposes to the cluster admins via the `openshift-monitoring/cluster-monitoring-operator` configmap (for "historical" reasons, CMO doesn't expose CRDs for this). It allows to configure both the operator (for example to enable user-workload monitoring) and the operands (for example to set resource requests/limits on Prometheus).